src/Microsoft.Extensions.Configuration.AzureAppConfiguration/OfflineFileCache.cs

```diff
@@ -237,75 +197,76 @@ private async Task DoUpdate(string tempFile, int retry = 0)
             }
         }
 
-        private void EnsureOptions(AzureAppConfigurationOptions azconfigOptions)
+        private void EnsureOptions(AzureAppConfigurationOptions appConfigOptions)
         {
-            if (azconfigOptions == null)
-            {
-                throw new ArgumentNullException(nameof(azconfigOptions));
-            }
-
-            if (azconfigOptions.ConnectionString == null)
-            {
-                throw new InvalidOperationException("An Azure App Configuration connection string is required.");
-            }
-
-            OfflineFileCacheOptions options = _options ?? new OfflineFileCacheOptions();
-
-            if ((options.Key == null) || (options.SignKey == null) || (options.IV == null))
+            if (string.IsNullOrEmpty(_appConfigEndpoint))
             {
-                byte[] secret = Convert.FromBase64String(ConnectionStringParser.Parse(azconfigOptions.ConnectionString, "Secret"));
-                using (SHA256 sha256 = SHA256.Create())
-                {
-                    byte[] hash = sha256.ComputeHash(secret);
-
-                    options.Key = options.Key ?? hash;
-                    options.SignKey = options.SignKey ?? hash;
-                    options.IV = options.IV ?? hash.Take(16).ToArray();
-                }
+                _appConfigEndpoint = (appConfigOptions.Endpoint != null)
+                                      ? appConfigOptions.Endpoint.ToString()
+                                      : ConnectionStringParser.Parse(appConfigOptions.ConnectionString, "Endpoint");
             }
-
+            
             if (string.IsNullOrEmpty(_scopeToken))
             {
-                //
                 // The default scope token is the configuration store endpoint combined with all of the key-value filters
+                var sb = new StringBuilder($"{_appConfigEndpoint}\0");
 
-                string endpoint = ConnectionStringParser.Parse(azconfigOptions.ConnectionString, "Endpoint");
-
-                if (string.IsNullOrWhiteSpace(endpoint))
-                {
-                    throw new InvalidOperationException("Invalid connection string format.");
-                }
-
-                var sb = new StringBuilder($"{endpoint}\0");
-
-                foreach (var selector in azconfigOptions.KeyValueSelectors)
+                foreach (var selector in appConfigOptions.KeyValueSelectors)
                 {
                     sb.Append($"{selector.KeyFilter}\0{selector.LabelFilter}\0");
                 }
 
                 _scopeToken = sb.ToString();
             }
 
-            _options = options;
+            if (_options.DataProtector == null)
+            {
+                IDataProtectionProvider dataProtectionProvider = DataProtectionProvider.Create($"{_options.Path}-{_appConfigEndpoint}");
+                _options.DataProtector = dataProtectionProvider.CreateProtector($"AppConfigurationOfflineFileCacheProtector-{_appConfigEndpoint}");

```

Comment:
>_options.DataProtector [](http://example.com/codeflow?start=16&length=22)

Do we need to set this? It kind of looks like it's just being used for conditional initialization logic.

Reply:
We do need to set `_options.DataProtector`. This is because every time we call `EnsureOptions`, we need to check if `_options.DataProtector` is null. If it's null, that means users have not provided an instance of DataProtector and we need to create a default instance only once.

Reply:
But if we check if _timeLimitedDataProtector is null instead don't we avoid that skip without assigning that on the input options?

Reply:
_timeLimitedDataProtector will be null on the first import/export call regardless of whether users have provided a DataProtector or not.

We will have to do something like this:
```
if (_timeLimitedDataProtector == null)
{
    if (_options.DataProtector == null)
    {
        IDataProtectionProvider dataProtectionProvider = DataProtectionProvider.Create($"{_options.Path}-{_appConfigEndpoint}");
       _timeLimitedDataProtector = dataProtectionProvider.CreateProtector($"AppConfigurationOfflineFileCacheProtector-{_appConfigEndpoint}").ToTimeLimitedDataProtector();
    }
    else
    {
        _timeLimitedDataProtector  = _options.DataProtector.ToTimeLimitedDataProtector();
    }
}
```

Reply:
Right, which makes sense if `_timeLimitedDataProtector` is what we are trying to initialize

Reply:
Okay, I'll change it to initialize _timeLimitedDataProtector  only.